// @mobilewright/driver-mobilecli/dist/rpc-client.js
import WebSocket from 'ws';
export class RpcError extends Error {
    constructor(message, code, data) {
        super(message);
        this.code = code;
        this.data = data;
        this.name = 'RpcError';
    }
}
export class RpcClient {
    constructor(url, requestTimeout = 30_000) {
        this.url = url;
        this.requestTimeout = requestTimeout;
        this.ws = null;
        this.nextId = 1;
        this.pending = new Map();
        this.connectionPromise = null;
    }
    async connect() {
        if (this.ws?.readyState === WebSocket.OPEN)
            return;
        if (this.connectionPromise)
            return this.connectionPromise;
        this.connectionPromise = new Promise((resolve, reject) => {
            let settled = false;
            const ws = new WebSocket(this.url);
            const timeout = setTimeout(() => {
                ws.terminate();
                if (!settled) {
                    settled = true;
                    reject(new Error(`Connection to ${this.url} timed out`));
                }
            }, this.requestTimeout);
            ws.on('open', () => {
                clearTimeout(timeout);
                this.ws = ws;
                this.connectionPromise = null;
                settled = true;
                resolve();
            });
            ws.on('error', (err) => {
                clearTimeout(timeout);
                this.connectionPromise = null;
                if (!settled) {
                    settled = true;
                    if (err instanceof AggregateError && err.errors.length > 0) {
                        reject(new Error(`Failed to connect to ${this.url}: ${err.errors.map((e) => e.message).join('; ')}`));
                    }
                    else {
                        reject(new Error(`Failed to connect to ${this.url}: ${err.message}`));
                    }
                }
            });
            ws.on('message', (data) => {
                this.handleMessage(data);
            });
            ws.on('close', (code, reason) => {
                this.ws = null;
                const reasonStr = reason.toString() || 'no reason';
                const msg = `WebSocket connection closed (code=${code}, reason=${reasonStr})`;
                if (!settled) {
                    clearTimeout(timeout);
                    this.connectionPromise = null;
                    settled = true;
                    reject(new Error(msg));
                }
                for (const [id, req] of this.pending) {
                    clearTimeout(req.timer);
                    req.reject(new Error(msg));
                    this.pending.delete(id);
                }
            });
        });
        return this.connectionPromise;
    }
    async call(method, params) {
        await this.connect();
        const ws = this.ws;
        if (!ws || ws.readyState !== WebSocket.OPEN) {
            throw new Error('WebSocket is not connected');
        }
        const id = this.nextId++;
        const request = {
            jsonrpc: '2.0',
            id,
            method,
            ...(params !== undefined && { params }),
        };
        return new Promise((resolve, reject) => {
            const timer = setTimeout(() => {
                this.pending.delete(id);
                reject(new Error(`RPC call "${method}" timed out after ${this.requestTimeout}ms`));
            }, this.requestTimeout);
            this.pending.set(id, {
                resolve: resolve,
                reject,
                timer,
            });
            ws.send(JSON.stringify(request));
        });
    }
    async disconnect() {
        if (this.ws) {
            this.ws.close();
            this.ws = null;
        }
    }
    get isConnected() {
        return this.ws?.readyState === WebSocket.OPEN;
    }
    handleMessage(data) {
        let response;
        try {
            response = JSON.parse(data.toString());
        }
        catch {
            return; // Ignore malformed messages
        }
        const pending = this.pending.get(response.id);
        if (!pending)
            return;
        this.pending.delete(response.id);
        clearTimeout(pending.timer);
        if (response.error) {
            // Surface the detail from mobilecli — response.error.data often
            // contains the real error (e.g. "exit status 4") while
            // response.error.message is just "Server error".
            const detail = typeof response.error.data === 'string'
                ? response.error.data
                : response.error.message;
            pending.reject(new RpcError(detail, response.error.code, response.error.data));
        }
        else {
            pending.resolve(response.result);
        }
    }
}
