// @mobilewright/core/dist/device.js
import { Screen } from './screen.js';
import { retryUntil } from './poll.js';
const LAUNCH_APP_TIMEOUT = 20_000;
export class Device {
    constructor(driver, opts = {}) {
        this.cleanupCallbacks = [];
        this._screen = null;
        this.driver = driver;
        this.opts = opts;
    }
    /** Register a callback to run on close(). Used by launchers for cleanup. */
    onClose(callback) {
        this.cleanupCallbacks.push(callback);
    }
    // ─── Connection lifecycle ────────────────────────────────────
    async connect(config) {
        return this.driver.connect(config);
    }
    async disconnect() {
        await this.driver.disconnect();
    }
    /** Full cleanup: disconnect + run any registered cleanup callbacks. */
    async close() {
        await this.disconnect();
        for (const cb of this.cleanupCallbacks) {
            await cb();
        }
        this.cleanupCallbacks = [];
    }
    get screen() {
        this._screen ??= new Screen(this.driver, this.opts.locatorDefaults);
        return this._screen;
    }
    // ─── Device control ──────────────────────────────────────────
    async getOrientation() {
        return this.driver.getOrientation();
    }
    async setOrientation(orientation) {
        return this.driver.setOrientation(orientation);
    }
    async openUrl(url) {
        return this.driver.openUrl(url);
    }
    /** Alias for openUrl — matches Playwright's page.goto(). */
    async goto(url) {
        return this.openUrl(url);
    }
    // ─── App control ─────────────────────────────────────────────
    async launchApp(bundleId, opts) {
        await this.driver.launchApp(bundleId, opts);
        if (opts?.noWaitAfter) {
            return;
        }
        try {
            await retryUntil(() => this.getForegroundApp(), (app) => app.bundleId === bundleId, LAUNCH_APP_TIMEOUT, `launchApp: timed out waiting for "${bundleId}" to be in foreground`);
        }
        catch (err) {
            if (String(err).includes('could not determine foreground app')) {
                // mobilecli's WebSocket RPC path for device.apps.foreground fails on
                // some Android devices even though the app launched successfully.
                // Warn and continue rather than failing the launch entirely.
                console.warn(`[mobilewright] warning: could not verify "${bundleId}" reached foreground — proceeding anyway. This is a known mobilecli issue on some Android devices.`);
                return;
            }
            throw err;
        }
    }
    async terminateApp(bundleId) {
        return this.driver.terminateApp(bundleId);
    }
    async listApps() {
        return this.driver.listApps();
    }
    async getForegroundApp() {
        return this.driver.getForegroundApp();
    }
    async installApp(path) {
        return this.driver.installApp(path);
    }
    async uninstallApp(bundleId) {
        return this.driver.uninstallApp(bundleId);
    }
    // ─── Recording ─────────────────────────────────────────────────
    async startRecording(opts) {
        return this.driver.startRecording(opts);
    }
    async stopRecording() {
        return this.driver.stopRecording();
    }
}
