// @mobilewright/core/dist/expect.js
import { LocatorError } from './locator.js';
import { retryUntil } from './poll.js';
import { filterStack } from './stackTrace.js';
const DEFAULT_TIMEOUT = 5_000;
export function expect(actual) {
    if (actual && typeof actual === 'object' && 'tap' in actual && 'getText' in actual) {
        return new LocatorAssertions(actual, false);
    }
    return new ValueAssertions(actual, false);
}
class LocatorAssertions {
    constructor(locator, negated) {
        this.locator = locator;
        this.negated = negated;
    }
    get not() {
        return new LocatorAssertions(this.locator, !this.negated);
    }
    async toBeVisible(opts) {
        await this.assertBoolean('visible', () => this.locator.isVisible({ timeout: 0 }), opts);
    }
    async toBeHidden(opts) {
        await this.assertBoolean('hidden', async () => {
            const visible = await this.locator.isVisible({ timeout: 0 });
            return !visible;
        }, opts);
    }
    async toBeEnabled(opts) {
        await this.assertBoolean('enabled', () => this.locator.isEnabled({ timeout: 0 }), opts);
    }
    async toBeDisabled(opts) {
        await this.assertBoolean('disabled', async () => {
            const enabled = await this.locator.isEnabled({ timeout: 0 });
            return !enabled;
        }, opts);
    }
    async toBeSelected(opts) {
        await this.assertBoolean('selected', () => this.locator.isSelected({ timeout: 0 }), opts);
    }
    async toBeFocused(opts) {
        await this.assertBoolean('focused', () => this.locator.isFocused({ timeout: 0 }), opts);
    }
    async toBeChecked(opts) {
        await this.assertBoolean('checked', () => this.locator.isChecked({ timeout: 0 }), opts);
    }
    async toHaveText(expected, opts) {
        await this.assertText((text) => expected instanceof RegExp ? expected.test(text) : text === expected, expected, opts);
    }
    async toContainText(expected, opts) {
        await this.assertText((text) => text.includes(expected), expected, opts);
    }
    async toHaveCount(expected, opts) {
        let lastCount = 0;
        await this.retryAssertion(async () => { lastCount = await this.locator.count(); return lastCount; }, (count) => {
            const matches = count === expected;
            return this.negated ? !matches : matches;
        }, opts?.timeout ?? DEFAULT_TIMEOUT, () => this.negated
            ? `Expected element count NOT to be ${expected}, but got ${lastCount}`
            : `Expected element count to be ${expected}, but got ${lastCount}`);
    }
    async toBeEmpty(opts) {
        let lastValue = '';
        await this.retryAssertion(async () => {
            try {
                lastValue = await this.locator.getValue({ timeout: 0 });
                return lastValue;
            }
            catch (e) {
                if (!(e instanceof LocatorError)) {
                    throw e;
                }
                return null;
            }
        }, (value) => {
            if (value === null) {
                return false;
            }
            const isEmpty = value === '';
            return this.negated ? !isEmpty : isEmpty;
        }, opts?.timeout ?? DEFAULT_TIMEOUT, () => this.negated
            ? 'Expected element NOT to be empty, but it was'
            : `Expected element to be empty, but got "${lastValue}"`);
    }
    async toHaveValue(expected, opts) {
        let lastValue = '';
        await this.retryAssertion(async () => {
            try {
                lastValue = await this.locator.getValue({ timeout: 0 });
                return lastValue;
            }
            catch (e) {
                if (!(e instanceof LocatorError)) {
                    throw e;
                }
                return null;
            }
        }, (value) => {
            if (value === null) {
                return false;
            }
            const matches = expected instanceof RegExp ? expected.test(value) : value === expected;
            return this.negated ? !matches : matches;
        }, opts?.timeout ?? DEFAULT_TIMEOUT, () => this.negated
            ? `Expected element NOT to have value "${expected}", but got "${lastValue}"`
            : `Expected element to have value "${expected}", but got "${lastValue}"`);
    }
    async assertBoolean(name, poll, opts) {
        await this.retryAssertion(poll, (result) => (this.negated ? !result : result), opts?.timeout ?? DEFAULT_TIMEOUT, this.negated
            ? `Expected element to NOT be ${name}, but it was`
            : `Expected element to be ${name}, but it was not`);
    }
    async assertText(predicate, expected, opts) {
        let lastText = '';
        await this.retryAssertion(async () => {
            try {
                lastText = await this.locator.getText({ timeout: 0 });
                return lastText;
            }
            catch (e) {
                if (!(e instanceof LocatorError)) {
                    throw e;
                }
                return null;
            }
        }, (text) => {
            if (text === null) {
                return false;
            }
            const matches = predicate(text);
            return this.negated ? !matches : matches;
        }, opts?.timeout ?? DEFAULT_TIMEOUT, () => this.negated
            ? `Expected element NOT to have text "${expected}", but got "${lastText}"`
            : `Expected element to have text "${expected}", but got "${lastText}"`);
    }
    async retryAssertion(poll, predicate, timeout, failMessage) {
        try {
            await retryUntil(poll, predicate, timeout, failMessage);
        }
        catch (e) {
            throw new ExpectError(e instanceof Error ? e.message : String(e));
        }
    }
}
class ValueAssertions {
    constructor(actual, negated) {
        this.actual = actual;
        this.negated = negated;
    }
    get not() {
        return new ValueAssertions(this.actual, !this.negated);
    }
    toBe(expected) {
        const pass = Object.is(this.actual, expected);
        this.assert(pass, `Expected ${fmt(expected)}, but received ${fmt(this.actual)}`);
    }
    toEqual(expected) {
        const pass = JSON.stringify(this.actual) === JSON.stringify(expected);
        this.assert(pass, `Expected ${fmt(expected)}, but received ${fmt(this.actual)}`);
    }
    toBeTruthy() {
        this.assert(!!this.actual, `Expected truthy, but received ${fmt(this.actual)}`);
    }
    toBeFalsy() {
        this.assert(!this.actual, `Expected falsy, but received ${fmt(this.actual)}`);
    }
    toBeGreaterThan(expected) {
        this.assert(this.actual > expected, `Expected ${fmt(this.actual)} > ${expected}`);
    }
    toBeLessThan(expected) {
        this.assert(this.actual < expected, `Expected ${fmt(this.actual)} < ${expected}`);
    }
    toBeCloseTo(expected, precision = 2) {
        const tolerance = Math.pow(10, -precision) / 2;
        const pass = Math.abs(this.actual - expected) < tolerance;
        this.assert(pass, `Expected ${fmt(this.actual)} to be close to ${expected} (precision ${precision})`);
    }
    toContain(expected) {
        const actual = this.actual;
        const pass = Array.isArray(actual)
            ? actual.includes(expected)
            : typeof actual === 'string' ? actual.includes(expected) : false;
        this.assert(pass, `Expected ${fmt(this.actual)} to contain ${fmt(expected)}`);
    }
    toBeNull() {
        this.assert(this.actual === null, `Expected null, but received ${fmt(this.actual)}`);
    }
    toBeUndefined() {
        this.assert(this.actual === undefined, `Expected undefined, but received ${fmt(this.actual)}`);
    }
    toMatch(pattern) {
        const str = String(this.actual);
        const regex = typeof pattern === 'string' ? new RegExp(pattern) : pattern;
        this.assert(regex.test(str), `Expected ${fmt(this.actual)} to match ${regex}`);
    }
    toBeInstanceOf(expected) {
        const pass = this.actual instanceof expected;
        this.assert(pass, `Expected instance of ${expected.name}, but received ${fmt(this.actual)}`);
    }
    toBeDefined() {
        this.assert(this.actual !== undefined, 'Expected defined, but received undefined');
    }
    toBeGreaterThanOrEqual(expected) {
        this.assert(this.actual >= expected, `Expected ${fmt(this.actual)} >= ${expected}`);
    }
    toBeLessThanOrEqual(expected) {
        this.assert(this.actual <= expected, `Expected ${fmt(this.actual)} <= ${expected}`);
    }
    toBeNaN() {
        this.assert(Number.isNaN(this.actual), `Expected NaN, but received ${fmt(this.actual)}`);
    }
    toContainEqual(expected) {
        const actual = this.actual;
        const pass = Array.isArray(actual) && actual.some((item) => JSON.stringify(item) === JSON.stringify(expected));
        this.assert(pass, `Expected ${fmt(this.actual)} to contain equal ${fmt(expected)}`);
    }
    toHaveLength(expected) {
        const actual = this.actual;
        const length = actual?.length ?? 0;
        this.assert(length === expected, `Expected length ${expected}, but received ${length}`);
    }
    toHaveProperty(key, value) {
        const actual = this.actual;
        const hasKey = actual != null && key in actual;
        const pass = value === undefined ? hasKey : hasKey && Object.is(actual[key], value);
        this.assert(pass, `Expected ${fmt(this.actual)} to have property "${key}"${value !== undefined ? ` with value ${fmt(value)}` : ''}`);
    }
    toMatchObject(expected) {
        const actual = this.actual;
        const pass = actual != null && Object.keys(expected).every((key) => JSON.stringify(actual[key]) === JSON.stringify(expected[key]));
        this.assert(pass, `Expected ${fmt(this.actual)} to match object ${fmt(expected)}`);
    }
    toStrictEqual(expected) {
        const pass = JSON.stringify(this.actual) === JSON.stringify(expected)
            && Object.getPrototypeOf(this.actual) === Object.getPrototypeOf(expected);
        this.assert(pass, `Expected ${fmt(expected)}, but received ${fmt(this.actual)}`);
    }
    toThrow(expected) {
        if (typeof this.actual !== 'function') {
            throw new ExpectError(`Expected a function, but received ${fmt(this.actual)}`);
        }
        const fn = this.actual;
        let threw = false;
        let error;
        try {
            fn();
        }
        catch (e) {
            threw = true;
            error = e;
        }
        if (expected === undefined) {
            this.assert(threw, 'Expected function to throw');
        }
        else {
            const message = threw && error instanceof Error ? error.message : String(error);
            const matches = typeof expected === 'string' ? message.includes(expected) : expected.test(message);
            this.assert(threw && matches, `Expected function to throw matching ${fmt(expected)}, but got ${fmt(message)}`);
        }
    }
    assert(pass, message) {
        const ok = this.negated ? !pass : pass;
        if (!ok) {
            throw new ExpectError(this.negated ? `Negation failed: ${message}` : message);
        }
    }
}
function fmt(value) {
    return typeof value === 'string' ? `"${value}"` : String(value);
}
export class ExpectError extends Error {
    constructor(message) {
        super(message);
        this.name = 'ExpectError';
        this.stack = filterStack(this.stack);
    }
}
