// @mobilewright/core/dist/locator.js
import sharp from 'sharp';
import { queryAll } from './query-engine.js';
import { sleep } from './sleep.js';
const DEFAULT_TIMEOUT = 5_000;
const DEFAULT_POLL_INTERVAL = 100;
const DEFAULT_STABILITY_DELAY = 50;
export class Locator {
    /** Create a root locator that searches the entire view hierarchy. */
    static root(driver, options = {}) {
        return new Locator(driver, { kind: 'root' }, options);
    }
    constructor(driver, strategy, options = {}) {
        this.driver = driver;
        this.strategy = strategy;
        this.options = options;
    }
    // ─── Chaining ────────────────────────────────────────────────
    getByLabel(label, opts) {
        return this.child({ kind: 'label', value: label, exact: opts?.exact });
    }
    getByTestId(testId) {
        return this.child({ kind: 'testId', value: testId });
    }
    getByText(text, opts) {
        return this.child({ kind: 'text', value: text, exact: opts?.exact });
    }
    getByType(type) {
        return this.child({ kind: 'type', value: type });
    }
    getByRole(role, opts) {
        return this.child({ kind: 'role', value: role, name: opts?.name });
    }
    getByPlaceholder(placeholder, opts) {
        return this.child({ kind: 'placeholder', value: placeholder, exact: opts?.exact });
    }
    child(childStrategy) {
        return new Locator(this.driver, { kind: 'chain', parent: this.strategy, child: childStrategy }, this.options);
    }
    // ─── Collection ──────────────────────────────────────────────
    first() {
        return this.nth(0);
    }
    last() {
        return this.nth(-1);
    }
    nth(index) {
        return new Locator(this.driver, { kind: 'nth', parent: this.strategy, index }, this.options);
    }
    async count() {
        const roots = await this.driver.getViewHierarchy();
        return queryAll(roots, this.strategy).length;
    }
    async all() {
        const roots = await this.driver.getViewHierarchy();
        const matches = queryAll(roots, this.strategy);
        return matches.map((_, i) => new Locator(this.driver, { kind: 'nth', parent: this.strategy, index: i }, this.options));
    }
    // ─── Actions ─────────────────────────────────────────────────
    async tap(opts) {
        const node = await this.resolveActionable(opts?.timeout);
        const { x, y } = centerOf(node.bounds);
        await this.driver.tap(x, y);
    }
    async doubleTap(opts) {
        const node = await this.resolveActionable(opts?.timeout);
        const { x, y } = centerOf(node.bounds);
        await this.driver.doubleTap(x, y);
    }
    async longPress(opts) {
        const node = await this.resolveActionable(opts?.timeout);
        const { x, y } = centerOf(node.bounds);
        await this.driver.longPress(x, y, opts?.duration);
    }
    async fill(text, opts) {
        const node = await this.resolveActionable(opts?.timeout);
        const { x, y } = centerOf(node.bounds);
        // Tap to focus, then type
        await this.driver.tap(x, y);
        await this.driver.typeText(text);
    }
    async screenshot(opts) {
        const node = await this.resolveVisible(opts?.timeout);
        const fullScreenshot = await this.driver.screenshot();
        return cropToElement(fullScreenshot, node.bounds, await this.driver.getScreenSize());
    }
    async swipe(opts) {
        const node = await this.resolveActionable(opts.timeout);
        const { x, y } = centerOf(node.bounds);
        await this.driver.swipe(opts.direction, { startX: x, startY: y });
    }
    async scrollIntoViewIfNeeded(opts) {
        const maxSwipes = opts?.maxSwipes ?? 10;
        const direction = opts?.direction ?? 'up';
        const screenSize = await this.driver.getScreenSize();
        const POST_SWIPE_SETTLE = 200;
        for (let i = 0; i < maxSwipes; i++) {
            const roots = await this.driver.getViewHierarchy();
            const node = queryAll(roots, this.strategy)[0] ?? null;
            if (node && isWithinViewport(node.bounds, screenSize)) {
                return;
            }
            const swipeDirection = node ? swipeDirectionToReveal(node.bounds, screenSize) : direction;
            await this.driver.swipe(swipeDirection);
            await sleep(POST_SWIPE_SETTLE);
        }
        throw new LocatorError(`Element not scrolled into view after ${maxSwipes} swipes`, this.strategy);
    }
    // ─── Queries (with auto-wait for visibility) ─────────────────
    async exists() {
        const node = await this.resolve(0);
        return node !== null;
    }
    async isVisible(opts) {
        try {
            await this.waitFor({ state: 'visible', timeout: opts?.timeout ?? 0 });
            return true;
        }
        catch {
            return false;
        }
    }
    async isEnabled(opts) {
        const node = await this.resolve(opts?.timeout ?? 0);
        return node !== null && node.isEnabled;
    }
    async isSelected(opts) {
        const node = await this.resolve(opts?.timeout ?? 0);
        return node !== null && node.isSelected === true;
    }
    async isFocused(opts) {
        const node = await this.resolve(opts?.timeout ?? 0);
        return node !== null && node.isFocused === true;
    }
    async isChecked(opts) {
        const node = await this.resolve(opts?.timeout ?? 0);
        return node !== null && node.isChecked === true;
    }
    async boundingBox(opts) {
        const node = await this.resolveVisible(opts?.timeout);
        return { x: node.bounds.x, y: node.bounds.y, width: node.bounds.width, height: node.bounds.height };
    }
    async getText(opts) {
        const node = await this.resolveVisible(opts?.timeout);
        return node.text ?? node.label ?? node.value ?? '';
    }
    async getValue(opts) {
        const node = await this.resolveVisible(opts?.timeout);
        return node.value ?? '';
    }
    async waitFor(opts) {
        await this.pollUntilState(opts?.state ?? 'visible', opts?.timeout);
    }
    // ─── Internal resolution ─────────────────────────────────────
    /** Wait for a visible node and return it. Used by getText, screenshot. */
    async resolveVisible(timeout) {
        const node = await this.pollUntilState('visible', timeout);
        return node;
    }
    /** Poll until the given state is satisfied. Returns the matched node (or null for hidden/disabled). */
    async pollUntilState(state, timeout) {
        const effectiveTimeout = timeout ?? this.options.timeout ?? DEFAULT_TIMEOUT;
        const pollInterval = this.options.pollInterval ?? DEFAULT_POLL_INTERVAL;
        const deadline = Date.now() + effectiveTimeout;
        while (true) {
            const roots = await this.driver.getViewHierarchy();
            const node = queryAll(roots, this.strategy)[0] ?? null;
            if (checkState(node, state)) {
                return node;
            }
            if (Date.now() >= deadline) {
                throw new LocatorError(`Locator timed out waiting for state "${state}" after ${effectiveTimeout}ms`, this.strategy);
            }
            await sleep(pollInterval);
        }
    }
    /** Resolve to a single actionable node (visible, enabled, stable bounds) */
    async resolveActionable(timeout) {
        const effectiveTimeout = timeout ?? this.options.timeout ?? DEFAULT_TIMEOUT;
        const pollInterval = this.options.pollInterval ?? DEFAULT_POLL_INTERVAL;
        const stabilityDelay = this.options.stabilityDelay ?? DEFAULT_STABILITY_DELAY;
        const deadline = Date.now() + effectiveTimeout;
        let previousBounds = null;
        let lastReason = 'no matching element found';
        while (true) {
            const roots = await this.driver.getViewHierarchy();
            const node = queryAll(roots, this.strategy)[0];
            if (!node) {
                lastReason = 'no matching element found';
            }
            else if (!node.isVisible) {
                lastReason = 'element found but not visible';
            }
            else if (!node.isEnabled) {
                lastReason = 'element found but not enabled';
            }
            else {
                // Stability check: bounds haven't changed since last poll
                if (previousBounds && boundsEqual(previousBounds, node.bounds)) {
                    return node;
                }
                previousBounds = { ...node.bounds };
                if (Date.now() >= deadline)
                    return node; // accept without stability
                await sleep(stabilityDelay);
                continue;
            }
            if (Date.now() >= deadline) {
                throw new LocatorError(`Locator: ${lastReason} after ${effectiveTimeout}ms`, this.strategy);
            }
            await sleep(pollInterval);
        }
    }
    /** Resolve without waiting — returns null if not found */
    async resolve(timeout) {
        const deadline = Date.now() + timeout;
        const pollInterval = this.options.pollInterval ?? DEFAULT_POLL_INTERVAL;
        do {
            const roots = await this.driver.getViewHierarchy();
            const matches = queryAll(roots, this.strategy);
            if (matches.length > 0)
                return matches[0];
            if (timeout <= 0)
                return null;
            await sleep(pollInterval);
        } while (Date.now() < deadline);
        return null;
    }
}
async function cropToElement(screenshot, bounds, screenSize) {
    const metadata = await sharp(screenshot).metadata();
    const scale = (metadata.width ?? 1) / screenSize.width;
    return sharp(screenshot)
        .extract({
        left: Math.round(bounds.x * scale),
        top: Math.round(bounds.y * scale),
        width: Math.round(bounds.width * scale),
        height: Math.round(bounds.height * scale),
    })
        .toBuffer();
}
function centerOf(bounds) {
    return {
        x: Math.round(bounds.x + bounds.width / 2),
        y: Math.round(bounds.y + bounds.height / 2),
    };
}
function boundsEqual(a, b) {
    return (a.x === b.x &&
        a.y === b.y &&
        a.width === b.width &&
        a.height === b.height);
}
function checkState(node, state) {
    switch (state) {
        case 'visible':
            return node !== null && node.isVisible;
        case 'hidden':
            return node === null || !node.isVisible;
        case 'enabled':
            return node !== null && node.isEnabled;
        case 'disabled':
            return node !== null && !node.isEnabled;
    }
}
function isWithinViewport(bounds, screen) {
    return bounds.y >= 0
        && bounds.y + bounds.height <= screen.height
        && bounds.x >= 0
        && bounds.x + bounds.width <= screen.width;
}
function swipeDirectionToReveal(bounds, screen) {
    const centerY = bounds.y + bounds.height / 2;
    // Element is below the viewport → swipe up to reveal it
    if (centerY > screen.height) {
        return 'up';
    }
    // Element is above the viewport → swipe down to reveal it
    return 'down';
}
export class LocatorError extends Error {
    constructor(message, strategy) {
        super(message);
        this.strategy = strategy;
        this.name = 'LocatorError';
    }
}
