// @mobilewright/core/dist/screen.js
import { writeFileSync, mkdirSync } from 'node:fs';
import { dirname } from 'node:path';
import { Locator } from './locator.js';
export class Screen {
    constructor(driver, locatorDefaults = {}) {
        this.driver = driver;
        this.root = Locator.root(driver, locatorDefaults);
    }
    // ─── Locator factories (delegated to root locator) ─────────
    getByLabel(label, opts) {
        return this.root.getByLabel(label, opts);
    }
    getByTestId(testId) {
        return this.root.getByTestId(testId);
    }
    getByText(text, opts) {
        return this.root.getByText(text, opts);
    }
    getByType(type) {
        return this.root.getByType(type);
    }
    getByRole(role, opts) {
        return this.root.getByRole(role, opts);
    }
    getByPlaceholder(placeholder, opts) {
        return this.root.getByPlaceholder(placeholder, opts);
    }
    // ─── Direct screen actions ──────────────────────────────────
    async screenshot(opts) {
        const buffer = await this.driver.screenshot(opts);
        if (opts?.path) {
            mkdirSync(dirname(opts.path), { recursive: true });
            writeFileSync(opts.path, buffer);
        }
        return buffer;
    }
    async swipe(direction, opts) {
        return this.driver.swipe(direction, opts);
    }
    async pressButton(button) {
        return this.driver.pressButton(button);
    }
    async tap(x, y) {
        return this.driver.tap(x, y);
    }
    async goBack() {
        return this.driver.pressButton('BACK');
    }
    // ─── View tree ──────────────────────────────────────────────────
    async viewTree() {
        return this.driver.getViewHierarchy();
    }
}
