// mobilewright/dist/config.js
import { access } from 'node:fs/promises';
import { isAbsolute, join, resolve } from 'node:path';
import { pathToFileURL } from 'node:url';
import { createRequire } from 'node:module';
const _require = createRequire(import.meta.url);
export function toArray(value) {
    if (value === undefined) {
        return [];
    }
    return Array.isArray(value) ? value : [value];
}
/** Type-safe config helper for mobilewright.config.ts files. */
export function defineConfig(config) {
    const ourSetup = _require.resolve('./device-pool/setup.js');
    const ourTeardown = _require.resolve('./device-pool/teardown.js');
    const userSetups = toArray(config.globalSetup);
    const userTeardowns = toArray(config.globalTeardown);
    return {
        workers: 1,
        ...config,
        globalSetup: userSetups.length > 0 ? [ourSetup, ...userSetups] : ourSetup,
        globalTeardown: userTeardowns.length > 0 ? [...userTeardowns, ourTeardown] : ourTeardown,
    };
}
const CONFIG_FILES = [
    'mobilewright.config.ts',
    'mobilewright.config.js',
    'mobilewright.config.mjs',
];
async function importConfig(fullPath) {
    const mod = await import(pathToFileURL(fullPath).href);
    let config = mod.default ?? mod;
    // Some loaders (e.g. Playwright's TS transpiler) double-wrap the default export
    if (config && typeof config === 'object' && 'default' in config) {
        config = config.default;
    }
    return config;
}
/**
 * Load mobilewright config.
 *
 * If `configFile` is provided, that file is loaded directly. Otherwise scans
 * `cwd` for mobilewright.config.{ts,js,mjs}. Returns empty config when nothing
 * is found.
 */
export async function loadConfig(cwd = process.cwd(), configFile) {
    if (configFile) {
        const fullPath = isAbsolute(configFile) ? configFile : resolve(cwd, configFile);
        return importConfig(fullPath);
    }
    for (const name of CONFIG_FILES) {
        const fullPath = join(cwd, name);
        try {
            await access(fullPath);
            return importConfig(fullPath);
        }
        catch {
            continue;
        }
    }
    return {};
}
