// mobilewright/dist/launchers.js
import { Device } from '@mobilewright/core';
import { MobilecliDriver, DEFAULT_URL } from '@mobilewright/driver-mobilecli';
import { MobileUseDriver } from '@mobilewright/driver-mobile-use';
import { ensureMobilecliReachable } from './server.js';
import { toArray } from './config.js';
export function createDriver(driverConfig, url) {
    if (driverConfig?.type === 'mobile-use') {
        return new MobileUseDriver({
            region: driverConfig.region,
            apiKey: driverConfig.apiKey,
        });
    }
    return new MobilecliDriver({ url });
}
export async function connectDevice(params) {
    // URL is baked into the driver at construction time; don't override it here.
    // Passing mobilecli's default URL into MobileUseDriver.connect() would send
    // requests to the wrong server.
    const driver = createDriver(params.driverConfig, params.url);
    const device = new Device(driver);
    await device.connect({
        platform: params.platform,
        deviceId: params.deviceId,
        timeout: params.timeout,
    });
    return device;
}
export async function installAndLaunchApps(device, opts) {
    const appsToInstall = toArray(opts.installApps);
    for (const appPath of appsToInstall) {
        await device.installApp(appPath);
    }
    if (opts.bundleId && opts.autoAppLaunch !== false) {
        await device.launchApp(opts.bundleId);
    }
}
export async function findDevice(params) {
    const url = params.url ?? DEFAULT_URL;
    const driver = createDriver(params.driverConfig, url);
    const devices = await driver.listDevices({ platform: params.platform });
    const match = devices
        .filter((d) => d.state === 'online')
        .filter((d) => !params.deviceId || d.id === params.deviceId)
        .filter((d) => !params.deviceName || params.deviceName.test(d.name))
        .at(0);
    if (!match) {
        throw new Error(`no online ${params.platform} device found`);
    }
    return match;
}
function createLauncher(platform) {
    return {
        async launch(opts = {}) {
            const driverConfig = opts.driver;
            const url = opts.url ?? DEFAULT_URL;
            let serverProcess;
            if (!driverConfig || driverConfig.type === 'mobilecli') {
                const ensured = await ensureMobilecliReachable(url, { autoStart: opts.autoStart ?? true });
                serverProcess = ensured.serverProcess ?? undefined;
            }
            const found = await findDevice({
                platform,
                deviceId: opts.deviceId,
                deviceName: opts.deviceName,
                driverConfig,
                url,
            });
            const device = await connectDevice({
                platform,
                deviceId: found.id,
                driverConfig,
                url,
                timeout: opts.timeout,
            });
            if (serverProcess) {
                const proc = serverProcess;
                device.onClose(() => Promise.resolve(proc.kill()).then(() => undefined));
            }
            await installAndLaunchApps(device, opts);
            return device;
        },
        async devices() {
            const driver = new MobilecliDriver();
            return driver.listDevices({ platform });
        },
    };
}
/** iOS platform launcher */
export const ios = createLauncher('ios');
/** Android platform launcher */
export const android = createLauncher('android');
