// mobilewright/dist/device-pool/domain/device-slot.js
export class DeviceSlotStateError extends Error {
    constructor(message) {
        super(message);
        this.name = 'DeviceSlotStateError';
    }
}
export class DeviceSlot {
    constructor() {
        this._state = 'allocating';
        this._installedApps = new Set();
    }
    get state() {
        return this._state;
    }
    get deviceId() {
        return this._deviceId;
    }
    get platform() {
        return this._platform;
    }
    get allocationId() {
        return this._allocationId;
    }
    get driver() {
        return this._driver;
    }
    get model() {
        return this._model;
    }
    get osVersion() {
        return this._osVersion;
    }
    get type() {
        return this._type;
    }
    markAvailable(deviceId, platform, driver, model, osVersion, type) {
        if (this._state !== 'allocating') {
            throw new DeviceSlotStateError(`markAvailable requires state 'allocating', got '${this._state}'`);
        }
        this._state = 'available';
        this._deviceId = deviceId;
        this._platform = platform;
        this._driver = driver;
        this._model = model;
        this._osVersion = osVersion;
        this._type = type;
    }
    claim(allocationId) {
        if (this._state !== 'available') {
            throw new DeviceSlotStateError(`claim requires state 'available', got '${this._state}'`);
        }
        this._state = 'allocated';
        this._allocationId = allocationId;
    }
    release() {
        if (this._state !== 'allocated') {
            throw new DeviceSlotStateError(`release requires state 'allocated', got '${this._state}'`);
        }
        this._state = 'available';
        this._allocationId = undefined;
    }
    recordAppInstalled(bundleId) {
        this._installedApps.add(bundleId);
    }
    isAppInstalled(bundleId) {
        return this._installedApps.has(bundleId);
    }
}
