// @mobilewright/driver-mobile-use/dist/driver.js
import { createReadStream } from 'node:fs';
import { stat } from 'node:fs/promises';
import { basename } from 'node:path';
import createDebug from 'debug';
import { RpcClient } from './rpc-client.js';
export const DEFAULT_URL = 'wss://api.mobilenexthq.com/ws';
const VALID_PLATFORMS = new Set(['ios', 'android']);
const VALID_DEVICE_TYPES = new Set(['real', 'simulator', 'emulator']);
const VALID_DEVICE_STATES = new Set(['online', 'offline']);
function toPlatform(value) {
    return VALID_PLATFORMS.has(value) ? value : undefined;
}
function toDeviceType(value) {
    return VALID_DEVICE_TYPES.has(value) ? value : 'real';
}
function toDeviceState(value) {
    return VALID_DEVICE_STATES.has(value) ? value : 'offline';
}
function elementToViewNode(el) {
    const bounds = el.rect ?? { x: 0, y: 0, width: 0, height: 0 };
    return {
        type: el.type ?? 'Unknown',
        label: el.label || undefined,
        identifier: el.identifier || el.name || undefined,
        value: el.value || undefined,
        text: el.text || undefined,
        placeholder: el.placeholder || undefined,
        isVisible: typeof el.visible === 'boolean' ? el.visible : bounds.width > 0 && bounds.height > 0,
        isEnabled: el.enabled ?? true,
        bounds,
        children: el.children?.map(elementToViewNode) ?? [],
        raw: { ...el },
    };
}
function appendQueryParam(url, key, value) {
    const separator = url.includes('?') ? '&' : '?';
    return `${url}${separator}${key}=${encodeURIComponent(value)}`;
}
function sanitizeFilename(name) {
    return name.replace(/[^0-9a-zA-Z_.]/g, '_');
}
const debug = createDebug('mw:driver-mobile-use');
export class MobileUseDriver {
    get deviceInfo() {
        if (!this.session) {
            return null;
        }
        return { model: this.session.model, osVersion: this.session.osVersion, type: this.session.type };
    }
    constructor(options = {}) {
        this.session = null;
        this.ownsLease = false;
        this.options = options;
    }
    // ─── Connection ──────────────────────────────────────────────
    async connect(config) {
        const baseUrl = config.url ?? DEFAULT_URL;
        const url = this.options.apiKey
            ? appendQueryParam(baseUrl, 'token', this.options.apiKey)
            : baseUrl;
        debug('connecting to %s', baseUrl);
        const rpc = new RpcClient(url, config.timeout);
        await rpc.connect();
        debug('websocket connected');
        const platform = config.platform;
        // When deviceId is provided the device is already allocated by the pool
        // coordinator. Connect directly without fleet.allocate; this instance
        // does not own the lease and must not call fleet.release on disconnect.
        if (config.deviceId) {
            debug('reusing pre-allocated device %s', config.deviceId);
            this.ownsLease = false;
            this.session = { deviceId: config.deviceId, platform, rpc };
            return { deviceId: config.deviceId, platform };
        }
        this.ownsLease = true;
        const filters = this.buildFilters(config);
        debug('allocating device with filters %o', filters);
        const result = await rpc.call('fleet.allocate', { filters });
        let deviceId;
        let model;
        let osVersion;
        let type;
        if (result?.state === 'allocating' && result.sessionId) {
            debug('device is provisioning, waiting for allocation (session=%s)', result.sessionId);
            const device = await this.waitForAllocation(rpc, result.sessionId, config.timeout);
            debug('allocated device %s (session=%s, model=%s)', device.id, result.sessionId, device.model);
            deviceId = device.id;
            model = device.model;
            osVersion = device.version;
            type = toDeviceType(device.type);
        }
        else if (result?.device?.id) {
            debug('allocated device %s (session=%s, model=%s)', result.device.id, result.sessionId, result.device.model);
            deviceId = result.device.id;
            model = result.device.model;
            osVersion = result.device.version;
            type = toDeviceType(result.device.type);
        }
        else {
            throw new Error(`Device allocation failed: ${JSON.stringify(result)}`);
        }
        this.session = { deviceId, platform, rpc, model, osVersion, type };
        return { deviceId, platform };
    }
    async waitForAllocation(rpc, sessionId, timeout = 300_000) {
        const pollInterval = 5_000;
        const deadline = Date.now() + timeout;
        while (Date.now() < deadline) {
            await new Promise((resolve) => setTimeout(resolve, pollInterval));
            const elapsed = Math.round((timeout - (deadline - Date.now())) / 1000);
            debug('waiting for device allocation, session=%s (%ds elapsed)', sessionId, elapsed);
            const result = await rpc.call('devices.list', {});
            const device = result.devices.find((d) => d.provider?.sessionId === sessionId);
            if (!device) {
                continue;
            }
            if (device.state !== 'allocating') {
                return device;
            }
        }
        throw new Error(`Timed out waiting for device allocation after ${timeout / 1000}s (session=${sessionId})`);
    }
    async disconnect() {
        const session = this.requireSession();
        if (this.ownsLease) {
            debug('releasing device %s', session.deviceId);
            await session.rpc.call('fleet.release', { deviceId: session.deviceId });
        }
        await session.rpc.disconnect();
        this.session = null;
        this.ownsLease = false;
        debug('disconnected');
    }
    buildFilters(config) {
        const filters = [
            { attribute: 'platform', operator: 'EQUALS', value: config.platform },
        ];
        if (config.deviceName) {
            const name = typeof config.deviceName === 'string'
                ? config.deviceName
                : config.deviceName.source;
            filters.push({ attribute: 'name', operator: 'CONTAINS', value: name });
        }
        if (config.osVersion) {
            filters.push({ attribute: 'version', operator: 'EQUALS', value: config.osVersion });
        }
        return filters;
    }
    // ─── UI hierarchy ───────────────────────────────────────────
    async getViewHierarchy() {
        const result = await this.call('device.dump.ui');
        return result.elements.map(elementToViewNode);
    }
    // ─── Input ──────────────────────────────────────────────────
    async tap(x, y) {
        await this.call('device.io.tap', { x: Math.round(x), y: Math.round(y) });
    }
    async doubleTap(x, y) {
        await this.call('device.io.tap', { x, y });
        await this.call('device.io.tap', { x, y });
    }
    async longPress(x, y, duration) {
        await this.call('device.io.longpress', { x, y, ...(duration !== undefined && { duration }) });
    }
    async typeText(text) {
        await this.call('device.io.text', { text });
    }
    async swipe(direction, opts) {
        const screen = await this.getScreenSize();
        const centerX = screen.width / 2;
        const centerY = screen.height / 2;
        const startX = opts?.startX ?? centerX;
        const startY = opts?.startY ?? centerY;
        const isHorizontal = direction === 'left' || direction === 'right';
        const defaultDistance = (isHorizontal ? screen.width : screen.height) * 0.5;
        const distance = opts?.distance ?? defaultDistance;
        let endX = startX;
        let endY = startY;
        switch (direction) {
            case 'up':
                endY = startY - distance;
                break;
            case 'down':
                endY = startY + distance;
                break;
            case 'left':
                endX = startX - distance;
                break;
            case 'right':
                endX = startX + distance;
                break;
        }
        await this.call('device.io.swipe', {
            x1: Math.round(startX),
            y1: Math.round(startY),
            x2: Math.round(endX),
            y2: Math.round(endY),
            ...(opts?.duration !== undefined && { duration: opts.duration }),
        });
    }
    async gesture(gestures) {
        await this.call('device.io.gesture', { actions: gestures.pointers });
    }
    async pressButton(button) {
        await this.call('device.io.button', { button });
    }
    // ─── Screen ─────────────────────────────────────────────────
    async screenshot(opts) {
        const result = await this.call('device.screenshot', {
            ...(opts?.format && { format: opts.format }),
            ...(opts?.quality !== undefined && { quality: opts.quality }),
        });
        let b64 = result.data;
        const commaIdx = b64.indexOf(',');
        if (commaIdx !== -1) {
            b64 = b64.slice(commaIdx + 1);
        }
        return Buffer.from(b64, 'base64');
    }
    async getScreenSize() {
        const result = await this.call('device.info');
        const info = result.device;
        return info.screenSize ?? { width: info.screenWidth ?? 0, height: info.screenHeight ?? 0, scale: 1 };
    }
    async getOrientation() {
        const result = await this.call('device.io.orientation.get');
        return result.orientation === 'landscape' ? 'landscape' : 'portrait';
    }
    async setOrientation(orientation) {
        await this.call('device.io.orientation.set', { orientation });
    }
    // ─── Recording ──────────────────────────────────────────────
    async startRecording(opts) {
        await this.call('device.screenrecord', {
            output: opts.output,
            ...(opts.timeLimit !== undefined && { timeLimit: opts.timeLimit }),
        });
    }
    async stopRecording() {
        const result = await this.call('device.screenrecord.stop');
        if (result.url) {
            debug('download screen recording from %s', result.url.split('?')[0]);
        }
        return result;
    }
    // ─── Apps ───────────────────────────────────────────────────
    async launchApp(bundleId, opts) {
        await this.call('device.apps.launch', {
            bundleId,
            ...(opts?.locales && { locales: opts.locales }),
        });
    }
    async terminateApp(bundleId) {
        await this.call('device.apps.terminate', { bundleId });
    }
    async listApps() {
        // iOS returns a flat array, Android returns { apps: [...] }.
        const result = await this.call('device.apps.list');
        const apps = Array.isArray(result) ? result : result.apps;
        return apps.map((app) => ({
            bundleId: app.bundleId ?? app.packageName ?? '',
            name: app.appName,
            version: app.version,
        }));
    }
    async getForegroundApp() {
        const result = await this.call('device.apps.foreground');
        return {
            bundleId: result.bundleId ?? result.packageName ?? '',
            name: result.appName,
            version: result.version,
        };
    }
    async installApp(filePath) {
        const fileInfo = await stat(filePath);
        const filename = sanitizeFilename(basename(filePath));
        debug('creating upload for %s (%d bytes)', filename, fileInfo.size);
        const upload = await this.call('uploads.create', {
            filename,
            filesize: fileInfo.size,
        });
        debug('uploading %s to S3 (uploadId=%s)', filename, upload.uploadId);
        const body = createReadStream(filePath);
        const response = await fetch(upload.uploadUrl, {
            method: 'PUT',
            headers: {
                'Content-Type': 'application/octet-stream',
                'Content-Length': String(fileInfo.size),
            },
            body,
            duplex: 'half',
        });
        if (!response.ok) {
            throw new Error(`Upload failed with status ${response.status}`);
        }
        debug('upload complete, installing app (uploadId=%s)', upload.uploadId);
        await this.call('device.apps.install', { uploadId: upload.uploadId });
        debug('app installed successfully: %s', filename);
    }
    async uninstallApp(bundleId) {
        await this.call('device.apps.uninstall', { bundleId });
    }
    // ─── Device ─────────────────────────────────────────────────
    async listDevices(opts) {
        const result = await this.call('device.list');
        let devices = result.data.devices;
        if (opts?.platform) {
            devices = devices.filter((d) => d.platform === opts.platform);
        }
        if (opts?.state) {
            devices = devices.filter((d) => d.state === opts.state);
        }
        return devices
            .filter((d) => toPlatform(d.platform) !== undefined)
            .map((d) => ({
            id: d.id ?? d.udid ?? '',
            name: d.name,
            platform: toPlatform(d.platform),
            type: toDeviceType(d.type),
            state: toDeviceState(d.state),
            model: d.model,
            osVersion: d.version,
        }));
    }
    async openUrl(url) {
        await this.call('device.url', { url });
    }
    // ─── Helpers ────────────────────────────────────────────────
    call(method, params) {
        const session = this.requireSession();
        return session.rpc.call(method, { deviceId: session.deviceId, ...params });
    }
    requireSession() {
        if (!this.session) {
            throw new Error('No active session. Call connect() first.');
        }
        return this.session;
    }
}
