// mobilewright/dist/device-pool/adapters/http-server.js
import { createServer } from 'node:http';
export class DevicePoolHttpServer {
    constructor(options) {
        this.responsesByAllocationId = new Map();
        this.pool = options.pool;
        this.server = createServer((req, res) => this.handle(req, res));
    }
    async listen() {
        return new Promise((resolve, reject) => {
            this.server.once('error', reject);
            this.server.listen(0, '127.0.0.1', () => {
                const address = this.server.address();
                if (typeof address === 'string' || address === null) {
                    reject(new Error('expected AddressInfo'));
                    return;
                }
                resolve(address.port);
            });
        });
    }
    async close() {
        return new Promise((resolve) => {
            for (const res of this.responsesByAllocationId.values()) {
                res.end();
                res.socket?.destroy();
            }
            this.responsesByAllocationId.clear();
            this.server.close(() => resolve());
        });
    }
    async handle(req, res) {
        if (req.method !== 'POST') {
            res.statusCode = 405;
            res.end();
            return;
        }
        switch (req.url) {
            case '/allocate':
                await this.handleAllocate(req, res);
                return;
            case '/release':
                await this.handleRelease(req, res);
                return;
            case '/installed/is-installed':
                await this.handleIsAppInstalled(req, res);
                return;
            case '/installed/record':
                await this.handleRecordAppInstalled(req, res);
                return;
            case '/shutdown':
                await this.handleShutdown(req, res);
                return;
            default:
                res.statusCode = 404;
                res.end();
        }
    }
    async handleIsAppInstalled(req, res) {
        const body = await readJson(req);
        const installed = this.pool.isAppInstalled(body.allocationId, body.bundleId);
        res.statusCode = 200;
        res.end(JSON.stringify({ installed }));
    }
    async handleRecordAppInstalled(req, res) {
        const body = await readJson(req);
        this.pool.recordAppInstalled(body.allocationId, body.bundleId);
        res.statusCode = 200;
        res.end(JSON.stringify({ ok: true }));
    }
    async handleShutdown(_req, res) {
        await this.pool.shutdown();
        res.statusCode = 200;
        res.end(JSON.stringify({ ok: true }));
    }
    async handleAllocate(req, res) {
        const body = await readJson(req);
        let handle;
        try {
            handle = await this.pool.allocate(body.criteria ?? {});
        }
        catch (err) {
            res.statusCode = 503;
            res.end(JSON.stringify({ error: err.message }));
            return;
        }
        this.responsesByAllocationId.set(handle.allocationId, res);
        const onClose = () => {
            if (this.responsesByAllocationId.get(handle.allocationId) === res) {
                this.responsesByAllocationId.delete(handle.allocationId);
                void this.pool.release(handle.allocationId);
            }
        };
        res.on('close', onClose);
        res.statusCode = 200;
        res.setHeader('content-type', 'application/x-ndjson');
        res.write(JSON.stringify(handle) + '\n');
    }
    async handleRelease(req, res) {
        const body = await readJson(req);
        await this.pool.release(body.allocationId);
        const heldResponse = this.responsesByAllocationId.get(body.allocationId);
        if (heldResponse) {
            this.responsesByAllocationId.delete(body.allocationId);
            heldResponse.end();
        }
        res.statusCode = 200;
        res.end(JSON.stringify({ ok: true }));
    }
}
function readJson(req) {
    return new Promise((resolve, reject) => {
        const chunks = [];
        req.on('data', (chunk) => chunks.push(chunk));
        req.on('end', () => {
            try {
                const text = Buffer.concat(chunks).toString('utf8');
                resolve(text.length === 0 ? {} : JSON.parse(text));
            }
            catch (err) {
                reject(err);
            }
        });
        req.on('error', reject);
    });
}
