// mobilewright/dist/device-pool/adapters/http-client.js
import { Agent, request } from 'node:http';
export class HttpDevicePoolClient {
    constructor(options) {
        this.openAllocateRequests = new Map();
        this.baseUrl = options.baseUrl;
        this.agent = new Agent({ keepAlive: true });
    }
    allocate(criteria) {
        return new Promise((resolve, reject) => {
            const url = new URL('/allocate', this.baseUrl);
            const req = request({
                method: 'POST',
                host: url.hostname,
                port: url.port,
                path: url.pathname,
                agent: this.agent,
                headers: { 'content-type': 'application/json' },
            }, (res) => {
                if (res.statusCode !== 200) {
                    let buffer = '';
                    res.on('data', (chunk) => { buffer += chunk.toString('utf8'); });
                    res.on('end', () => reject(new Error(`HTTP ${res.statusCode}: ${buffer}`)));
                    return;
                }
                let buffer = '';
                const onData = (chunk) => {
                    buffer += chunk.toString('utf8');
                    const newlineIdx = buffer.indexOf('\n');
                    if (newlineIdx === -1) {
                        return;
                    }
                    const handle = JSON.parse(buffer.slice(0, newlineIdx));
                    res.off('data', onData);
                    this.openAllocateRequests.set(handle.allocationId, req);
                    // If the server closes the stream (e.g. on shutdown), release the map entry.
                    res.on('close', () => this.openAllocateRequests.delete(handle.allocationId));
                    resolve(handle);
                };
                res.on('data', onData);
                res.on('error', reject);
            });
            req.on('error', reject);
            req.write(JSON.stringify({ criteria }));
            req.end();
        });
    }
    async release(allocationId) {
        const openReq = this.openAllocateRequests.get(allocationId);
        if (openReq) {
            this.openAllocateRequests.delete(allocationId);
            openReq.destroy();
        }
        await this.postJson('/release', { allocationId });
    }
    async isAppInstalled(allocationId, bundleId) {
        const body = await this.postJson('/installed/is-installed', { allocationId, bundleId });
        return body.installed;
    }
    async recordAppInstalled(allocationId, bundleId) {
        await this.postJson('/installed/record', { allocationId, bundleId });
    }
    postJson(path, body) {
        return new Promise((resolve, reject) => {
            const url = new URL(path, this.baseUrl);
            const req = request({
                method: 'POST',
                host: url.hostname,
                port: url.port,
                path: url.pathname,
                agent: this.agent,
                headers: { 'content-type': 'application/json' },
            }, (res) => {
                const chunks = [];
                res.on('data', (c) => chunks.push(c));
                res.on('end', () => {
                    const text = Buffer.concat(chunks).toString('utf8');
                    if (res.statusCode !== 200) {
                        reject(new Error(`HTTP ${res.statusCode}: ${text}`));
                        return;
                    }
                    resolve(text.length === 0 ? {} : JSON.parse(text));
                });
                res.on('error', reject);
            });
            req.on('error', reject);
            req.write(JSON.stringify(body));
            req.end();
        });
    }
}
