// mobilewright/dist/device-pool/adapters/mobile-use-allocator.js
import createDebug from 'debug';
import { MobileUseDriver } from '@mobilewright/driver-mobile-use';
const debug = createDebug('mw:device-pool:mobile-use');
export class MobileUseAllocator {
    constructor(options) {
        this.activeDrivers = new Map();
        this.driverOptions = options.driverOptions;
    }
    async allocate(criteria) {
        debug('allocating device (criteria=%o)', criteria);
        const driver = new MobileUseDriver(this.driverOptions);
        const session = await driver.connect({
            platform: criteria.platform ?? 'ios',
            deviceName: criteria.deviceNamePattern ? new RegExp(criteria.deviceNamePattern) : undefined,
            deviceId: criteria.deviceId,
        });
        this.activeDrivers.set(session.deviceId, driver);
        debug('allocated device %s (platform=%s)', session.deviceId, session.platform);
        const info = driver.deviceInfo;
        return { deviceId: session.deviceId, platform: session.platform, driver: 'mobile-use', model: info?.model, osVersion: info?.osVersion, type: info?.type };
    }
    async release(deviceId) {
        debug('releasing device %s', deviceId);
        const driver = this.activeDrivers.get(deviceId);
        if (driver) {
            this.activeDrivers.delete(deviceId);
            await driver.disconnect();
            debug('released device %s', deviceId);
        }
    }
}
